### PR TITLE
Ingest parallaxes from Ultracool sheet

### LINKED ISSUE #527: Ingest literature parallaxes from Ultracool Sheet
Part of #452 

Ingest literature parallaxes. Columns ED:EF

Make sure adopted column is handled well.

### PR DESCRIPTION
**Pull request by Exu-112**
Short description: Ingest of parallaxes from Ultracool sheet, as well as rework of ingest_parallax function

Link to relevant issue: Closes #527 

For data ingests:
- [x] includes script used for ingest
- [ ] includes modified JSON files
- [x] Add new tests
- [ ] Update the Versions table

### CODE REVIEW

File: simple/utils/astrometry.py [hunk 1]
```diff
@@ -181,6 +182,103 @@ def ingest_parallaxes(db, sources, plxs, plx_errs, plx_refs, comments=None):
...
+        "reference": reference,
+        "adopted": adopted,
+        "comments": comment,
+    }
+
+    plx_obj = Parallaxes(**parallax_data)
+    with db.session as session:
+        session.add(plx_obj)
+        session.commit()
+    logger.info(f" Photometry added to database: {parallax_data}\n")
```
Comment (kelle) [inline simple/utils/astrometry.py:279]: Put into try/except
```Python
try:
            ...
            n_added += 1
            logger.info(f"Parallax added to database: \n " f"{parallax_data}")
        except sqlalchemy.exc.IntegrityError as e:
            msg = (
                "The source may not exist in Sources table.\n"
                "The parallax reference may not exist in Publications table. "
                "Add it with add_publication function. \n"
                "The parallax measurement may be a duplicate."
            )
            logger.error(msg)
            raise AstroDBError(msg) from e
```

File: simple/utils/astrometry.py [hunk 2]
```diff
@@ -181,6 +182,103 @@ def ingest_parallaxes(db, sources, plxs, plx_errs, plx_refs, comments=None):
...
 
 
+def ingest_parallax(
+    db,
+    source: str = None,
+    parallax: float = None,
+    plx_error: float = None,
+    reference: str = None,
+    comment: str = None,
+):
```
Comment (kelle) [inline simple/utils/astrometry.py:192]: Add docstring.

File: simple/utils/astrometry.py [hunk 3]
```diff
@@ -181,6 +182,103 @@ def ingest_parallaxes(db, sources, plxs, plx_errs, plx_refs, comments=None):
     return
 
 
+def ingest_parallax(
+    db,
+    source: str = None,
+    parallax: float = None,
+    plx_error: float = None,
+    reference: str = None,
+    comment: str = None,
```
Comment (kelle) [inline simple/utils/astrometry.py:191]: add `raise_error` flag so user can decide if they want the script to fail or to keep going.
```python
if raise_error:
    raise
else:
    logger.warning
```

File: tests/test_astrometry.py [hunk 4]
```diff
@@ -68,25 +69,57 @@ def t_rv():
...
+    assert results["adopted"][0]  # 1st source with ref 2 should be adopted
+
+
+def test_parallax_exceptions(temp_db):
+    try:
+        ingest_parallax(temp_db, "bad source", 1, 1, "Ref 1")
+    except AstroDBError:
+        None
+    else:  # weird syntax that activates if there is NO exception
+        assert False, "ingest_parallax ingested bad source without exception"
```
Comment (kelle) [inline tests/test_astrometry.py:107]:
```python
with pytest.raises(AstroDBError) as error_message:
        ingest_radial_velocity(
            temp_db, source="not a source", rv=12.5, rv_err=0.5, reference="Ref 1"
        )
    assert "No unique source match" in str(error_message.value)
```

File: simple/utils/astrometry.py [hunk 5]
```diff
@@ -181,6 +182,135 @@ def ingest_parallaxes(db, sources, plxs, plx_errs, plx_refs, comments=None):
     return
 
 
+def ingest_parallax(
```
Comment (kelle) [inline simple/utils/astrometry.py:185]: You can go ahead and delete the old ingest_parallaxes function.

**Comment by kelle** [comment]
return a dictionary called `flags`.  
- "added": True if it's added and False if it's skipped. 
- "content": the data that was attempted to add
- "message": string which includes information about why skipped

**Comment by Exu-112** [comment]
> return a dictionary called `flags`.
> 
> * "added": True if it's added and False if it's skipped.
> * "content": the data that was attempted to add
> * "message": string which includes information about why skipped

How should errors be handled? Should we return with the flags immediately or raise an error with the flags contained within? The first option would be a little confusing with regard to the raise_error flag and the fact that we don't raise errors anymore, but the second doesn't give you the flags dictionary anyways which seems to defeat the point of having it in the first place.

**Comment by kelle** [comment]
When raise_error is True, you don't need to return anything. But when raise_error is False is when the flags are most important cause it should return added: False.

File: simple/utils/astrometry.py [hunk 6]
```diff
@@ -181,6 +182,135 @@ def ingest_parallaxes(db, sources, plxs, plx_errs, plx_refs, comments=None):
...
+    db: astrodbkit2.astrodb.Database
+        Database object
+    sources: str
+        source name
+    plxs: float
+        parallax corresponding to the source
+    plx_errs: float
+        parallax uncertainties
+    plx_refs: str
+        reference for the parallax data
```
Comment (kelle) [inline simple/utils/astrometry.py:207]: docstring should agree with keywords in the function

File: simple/utils/astrometry.py [hunk 7]
```diff
@@ -2,183 +2,167 @@
...
+    parallax: float
+        parallax corresponding to the source
+    plx_err: float
+        parallax uncertainties
+    reference: str
+        reference for the parallax data
+    comment: str
+        comments
+    raise_error: bool
+        raise error if there is an error during ingest
```
Comment (kelle) [inline simple/utils/astrometry.py:46]: Be explicit
True: raises an error
False: does not raise an error, returns flags dictionary.

File: simple/utils/astrometry.py [hunk 8]
```diff
@@ -2,183 +2,167 @@
...
 
     """
```
Comment (kelle) [inline simple/utils/astrometry.py:48]: Add `Returns` part of docstring.

File: scripts/ingests/ultracool_sheet/Ingest_Parallax.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        multiple_sources += 1
+
+
+# 1108 data points in UC sheet in total
+print(f"ingested:{ingested}")  # 1013 ingested
+print(f"already exists:{already_exists}")  # skipped 6 due to preexisting data
+print(f"no sources:{no_sources}")  # skipped 86 due to 0 matches
+print(f"multiple sources:{multiple_sources}")  # skipped 2 due to multiple matches
+print(f"total:{ingested+already_exists+no_sources+multiple_sources}")  # 1108
+if DB_SAVE:
```
Comment (kelle) [inline scripts/ingests/ultracool_sheet/Ingest_Parallax.py:105]: add a check which makes sure the counts add up to what you expect.

**Review by kelle — CHANGES_REQUESTED** [review]
Looking great but still some tweaking to do.

File: scripts/ingests/ultracool_sheet/Ingest_Parallax.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+    if len(match) == 1:
+        # 1 Match found. INGEST!
+        simple_source = match[0]
+        logger.debug(f"Match found for {uc_sheet_name}: {simple_source}")
+
+        try:
+            references = source["ref_plx_lit"].split(";")
+            if references[0] == "Harr15":  # weird reference in UC sheet.
+                reference = "Harr15"
```
Comment (kelle) [inline scripts/ingests/ultracool_sheet/Ingest_Parallax.py:83]: delete.

File: simple/utils/astrometry.py [hunk 11]
```diff
@@ -2,183 +2,167 @@
...
+    try:
+        plx_obj = Parallaxes(**parallax_data)
+        with db.session as session:
+            session.add(plx_obj)
+            session.commit()
+        logger.info(f" Photometry added to database: {parallax_data}\n")
+        flags["added"] = True
+        return flags
+    except sqlalchemy.exc.IntegrityError as error_msg:
+        msg = ""
```
Comment (kelle) [inline simple/utils/astrometry.py:145]: Just for readability  set `flags["added"] = False` here.

File: tests/test_astrometry.py [hunk 12]
```diff
@@ -68,25 +69,50 @@ def t_rv():
...
+    with pytest.raises(AstroDBError) as error_message:
+        ingest_parallax(temp_db, "Fake 1", 1, 1, "bad ref")
+    assert "does not exist in Publications table" in str(error_message.value)
+
+    ingest_parallax(temp_db, "Fake 2", 1, 1, "Ref 2")
+    with pytest.raises(AstroDBError) as error_message:
+        ingest_parallax(temp_db, "Fake 2", 1, 1, "Ref 2")
+    assert "Duplicate measurement exists with same reference" in str(
+        error_message.value
+    )
```
Comment (kelle) [inline tests/test_astrometry.py:115]: Duplicate these tests with `raise_error = False`.  In each scenario, test all three elements of the flags dictionary.

File: simple/utils/astrometry.py [hunk 13]
```diff
@@ -2,183 +2,177 @@
...
-# PARALLAXES
-def ingest_parallaxes(db, sources, plxs, plx_errs, plx_refs, comments=None):
+def ingest_parallax(
+    db,
+    source: str = None,
+    parallax: float = None,
+    plx_error: float = None,
+    reference: str = None,
+    comment: str = None,
+    raise_error: bool = True,
```
Comment (kelle) [inline simple/utils/astrometry.py:27]: I think these should be keyword arguments.

File: tests/test_astrometry.py [hunk 14]
```diff
@@ -68,25 +69,87 @@ def t_rv():
...
+        },
+        "message": "Source 'bad source' does not exist in Sources table. ",
+    }
+
+    with pytest.raises(AstroDBError) as error_message:
+        ingest_parallax(temp_db, "Fake 1", 1, 0, "bad ref")
+    assert "does not exist in Publications table" in str(error_message.value)
+
+    flags = ingest_parallax(temp_db, "Fake 1", 1, 0, "bad ref", "comment", False)
+    print(flags)
```
Comment (kelle) [inline tests/test_astrometry.py:125]: delete.

File: tests/test_astrometry.py [hunk 15]
```diff
@@ -68,25 +69,87 @@ def t_rv():
...
+
+    ingest_parallax(temp_db, "Fake 2", 1, 1, "Ref 2")
+    with pytest.raises(AstroDBError) as error_message:
+        ingest_parallax(temp_db, "Fake 2", 1, 1, "Ref 2")
+    assert "Duplicate measurement exists with same reference" in str(
+        error_message.value
+    )
+
+    flags = ingest_parallax(temp_db, "Fake 2", 1, 1, "Ref 2", "comment", False)
+    print(flags)
```
Comment (kelle) [inline tests/test_astrometry.py:147]: delete.

File: scripts/ingests/ultracool_sheet/Ingest_Parallax.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+            if len(references) > 1:
+                comment = f"other references: {uc_ref_to_simple_ref(db, references[1])}"
+            ingest_parallax(
+                db,
+                simple_source,
+                source["plx_lit"],
+                source["plxerr_lit"],
+                reference,
+                comment,
+            )
```
Comment (kelle) [inline scripts/ingests/ultracool_sheet/Ingest_Parallax.py:93]: use the keywords.

File: simple/utils/astrometry.py [hunk 17]
```diff
@@ -2,183 +2,177 @@
...
-        list of parallaxes uncertainties
-    plx_refs: str or list[str]
-        list of references for the parallax data
-    comments: Optional[Union[List[str], str]]
+    source: str
+        source name
+    parallax: float
+        parallax of source in arcseconds
+    plx_err: float
+        parallax uncertainty in arcseconds
```
Comment (kelle) [inline simple/utils/astrometry.py:40]: Units are milliarcseconds

Comment (kelle) [inline simple/utils/astrometry.py:40]:
```suggestion
    parallax_mas: float
        parallax of source in arcseconds
    parallax_err_mas: float
        parallax uncertainty in arcseconds
```

**Review by kelle — APPROVED** [review]

**Pull request merged**